deepgram/audio/microphone/microphone.py

```diff
@@ -69,80 +90,71 @@ def start(self):
         self.logger.info("channels: %d", self.channels)
         self.logger.info("rate: %d", self.rate)
         self.logger.info("chunk: %d", self.chunk)
+        self.logger.info("input_device_id: %d", self.input_device_index)
 
         self.stream = self.audio.open(
             format=self.format,
             channels=self.channels,
             rate=self.rate,
             input=True,
             frames_per_buffer=self.chunk,
+            input_device_index=self.input_device_index,
+            stream_callback=self._callback
         )
 
-        self.exit = False
-        self.lock = threading.Lock()
-
+        self.exit.clear()
         self.stream.start_stream()
-        self.thread = threading.Thread(target=self._processing)
-        self.thread.start()
 
         self.logger.notice("start succeeded")
         self.logger.debug("Microphone.start LEAVE")
 
-    def _processing(self):
+    def _callback(self, input_data, frame_count, time_info, status_flags):
         """
-        the main processing loop for the microphone
+        The callback used to process data in callback mode.
         """
-        self.logger.debug("Microphone._processing ENTER")
+        import pyaudio
+        self.logger.debug("Microphone._callback ENTER")
 
-        try:
-            while True:
-                data = self.stream.read(self.chunk)
-
-                self.lock.acquire()
-                localExit = self.exit
-                self.lock.release()
-                if localExit:
-                    self.logger.info("exit is True")
-                    break
-                if data is None:
-                    self.logger.info("data is None")
-                    continue
-
-                if inspect.iscoroutinefunction(self.push_callback):
-                    self.logger.verbose("async/await callback")
-                    asyncio.run(self.push_callback(data))
-                else:
-                    self.logger.verbose("regular threaded callback")
-                    self.push_callback(data)
-
-            self.logger.notice("_processing exiting...")
-            self.logger.debug("Microphone._processing LEAVE")
+        if self.exit.is_set():
+            self.logger.info("exit is Set")
+            self.logger.notice("_callback stopping...")
+            self.logger.debug("Microphone._callback STOP")
```

Comment:
The reason I used STOP rather than LEAVE was that these are two different scenarios.
We Enter/Leave the callback many times over as the audio is running, and I only placed STOPs where we're either returning paAbort or raising an exception (therefore stopping the microphone).

Reply:
The idea with ENTER/LEAVE is that someone can write a script to see if a function never exits in a multithreaded app. Looking for the pair to know something is wrong. Not having this would break some of the audit tools we have.